src/main/java/net/minestom/server/particle/ParticleOptions.java

```diff
@@ -0,0 +1,148 @@
+package net.minestom.server.particle;
+
+import net.minestom.server.color.Color;
+import net.minestom.server.coordinate.Point;
+import net.minestom.server.entity.Entity;
+import net.minestom.server.item.ItemStack;
+import net.minestom.server.utils.MathUtils;
+import net.minestom.server.utils.binary.BinaryWriter;
+import net.minestom.server.utils.validate.Check;
+import org.jetbrains.annotations.NotNull;
+
+import static net.minestom.server.particle.ParticleOptions.*;
+
+public sealed abstract class ParticleOptions permits BlockStateOptions, Dust, DustColorTransition, Item, SculkCharge, Shriek, Vibration {
+    protected final BinaryWriter writer = new BinaryWriter();
+    private final Particle type;
+
+    public ParticleOptions(Particle type) {
+        this.type = type;
+    }
+
+    public byte[] getData() {
+        return writer.toByteArray();
+    }
+
+    public Particle type() {
+        return type;
+    }
+    
+    /**
+     * Options for {@link Particle#DUST}
+     */
+    public static final class Dust extends ParticleOptions {
```

Comment:
All these options should be records, duplicating a bit of code here because of records is fine.